### ISSUE TITLE
MAINT: Unify error handling

### ISSUE THREAD

**Issue by larsoner** [openmeeg/openmeeg#541]
... also there are a lot of `exit(...)` in the code -- that's causing the silent exit I'm debugging in https://github.com/openmeeg/openmeeg/pull/540. I'm not sure the best way to handle this. Maybe these should all become thrown exceptions, then these should be handled at the app level?

@papadop it looks like you converted only some of the `PyErr_SetString` [in #5365] into `throw Error(...)`. Why not convert them all?

_Originally posted by @larsoner in https://github.com/openmeeg/openmeeg/issues/536#issuecomment-1215638783_

_Originally posted by @larsoner in https://github.com/openmeeg/openmeeg/issues/536#issuecomment-1215679749_

**Comment by agramfort** [comment]
I am not sure I follow but the python bindings should not make the interpreter crash for any reason. Just raise exceptions

**Comment by papadop** [comment]
I agree that library code should throw exceptions that should be caught at app level. Then I do not understand why there would be a silent error if the exit code is non-zero (in case of error). Then I just saw that there is an exit(0) in danielsson.cpp, which seems wrong. I doubt this is triggered with C++ code, but indeed these non throwing exits are most probably a problem for python. So indeed we should convert them all to exceptions.

An exception to the rule of converting exits into throw is maybe commandline.h which is exclusively used in apps (so not exposed to python). One issue that will arise from that is that we will start to duplicate some error handling code whereas with the current implementation of geometry.cpp the IOs exception handling was commonized. Maybe for that case, we might just want to protect the exit calls with a #ifndef SWIGPYTHON .

All this probably comes from history.

**Comment by papadop** [comment]
@larsoner for 2:
There is no reason, except that maybe I was trying to work around a problem I was seeing whereas I did not see problems with the other ones. I must say I do not perfectly understand things here. I initially thought that setting PyErr_SetString was handled properly to generate a python error. But then, I discovered that when you set PyErr_SetString in a subroutine that does not return a null pointer, then it's not handled properly. So I have mostly replaced the PyErr_SetString for internal routines that were not returning pointers (I think, but maybe I propagated the solution to other cases). For the other ones, I believe that PyErr_SetString works as expected (if we return a null pointer).

Now given that in swig we catch all exceptions and convert them to python exceptions, I agree that relying on exception is more homogeneous and probably better.

**Comment by larsoner** [comment]
> Then I do not understand why there would be a silent error if the exit code is non-zero (in case of error).

Calling `exit(1)` in things like `Geometry` result in the *python interpreter* exiting immediately ("crashing" as users will think of it). This is bad. So I think the solution is to only use `exit(...)` in the C++ apps, and use `throw` in the `OpenMEEG` / `OpenMEEGMaths` libraries.

**Comment by papadop** [comment]
Totally agreeing then....
